api/src/gmsa_service.cpp

```diff
@@ -126,7 +126,10 @@ class CredentialsFetcherImpl final
     {
         server_->Shutdown();
         // Always shutdown the completion queue after the server.
-        cq_->Shutdown();
+         cq_->Shutdown();
+        void* ignored_tag;
+        bool ignored_ok;
+        while (cq_->Next(&ignored_tag, &ignored_ok)) { }
```

Comment:
1. does this need an upper limit to prevent the possibility of the while loop spinning forever?
2.  should the body of the loop do a thread sleep?  Presumably, this is waiting for another thread to do something - the application might be more performant if you slept for a millisecond to yield the processor back to the other thread.

Reply:
It will not spin for ever because the systemd watchdog will kill the process. Also, since this is the shutdown path, there is no reason to delay the shutdown.